music21/opensheetmusicdisplay/osmd.py

```diff
@@ -0,0 +1,189 @@
+
+        xml = open(score.write('musicxml')).read()
+        script = self.musicXMLToScript(xml, div_id, offline=offline)
+
+        display(Javascript(script))
+        return div_id
+
+    @staticmethod
+    def getUniqueDivId():
+        return "OSMD-div-"+ \
+                str(random.randint(0,1000000))+ \
+                "-"+str(time.time()).replace('.','-') # '.' is the class selector
+
+    @staticmethod
+    def musicXMLToScript(xml, div_id, offline=True):
+        
+        # print('xml length:', len(xml))
+        script_dir = os.path.join(os.path.dirname(__file__), 'opensheetmusicdisplay.0.3.1.min.js')
+
+        # script that will replace div contents with OSMD display
+        script = """
+        console.log("loadOSMD()");
+        function loadOSMD() { 
+            return new Promise(function(resolve, reject){
+
+                if (window.opensheetmusicdisplay) {
+                    console.log("already loaded")
+                    return resolve(window.opensheetmusicdisplay)
+                }
+                console.log("loading osmd for the first time")
+                // OSMD script has a 'define' call which conflicts with requirejs
+                var _define = window.define // save the define object 
+                window.define = undefined // now the loaded script will ignore requirejs
+                var s = document.createElement( 'script' );
+                // alternative to local file is using a CDN
+                // s.setAttribute( 'src', "https://cdn.jsdelivr.net/npm/opensheetmusicdisplay@0.3.1/build/opensheetmusicdisplay.min.js" );
+                //s.setAttribute( 'src', "{{script_dir}}" );
+                function oncompleted(){
+                    window.define = _define
+                    console.log("loaded OSMD for the first time",opensheetmusicdisplay)
+                    resolve(opensheetmusicdisplay);
+                };
+                {{script_command}}
+                
+            }) 
+        }
+        loadOSMD().then((OSMD)=>{
+            console.log("loaded OSMD",OSMD)
+            var div_id = "{{DIV_ID}}";
+                console.log(div_id)
+            document.querySelector('#'+div_id).innerHTML = "";
+            window.openSheetMusicDisplay = new OSMD.OpenSheetMusicDisplay(div_id);
+            openSheetMusicDisplay
+                .load({{data}})
+                .then(
+                  function() {
+                    console.log("rendering data")
+                    openSheetMusicDisplay.render();
+                  }
+                );
+        })
+        """ \
+        .replace('{{DIV_ID}}',div_id) \
+        .replace('{{data}}',json.dumps(xml)) \
+        .replace('{{script_dir}}',pathname2url(script_dir))
+
+        if offline is True:
+            script_content = open(script_dir).read()
+            script = script.replace('{{script_command}}',"""
+                s.type = 'text/javascript';
+                s.text="""+json.dumps(script_content)+""";
```

Comment:
Can you explain this bit some more.  Is this secure against maliciously coded XML?  Can script_content be something that causes the json.dumps to produce text that leaks into the rest of the code.  I don't like injecting user content into a general script tag.